src/storage/page_buffer.c

```diff
@@ -8185,15 +8185,66 @@ pgbuf_read_page_from_file_or_page_server (THREAD_ENTRY * thread_p, const VPID *
 	  const size_t io_page_size = static_cast<size_t> (db_io_page_size ());
 	  std::unique_ptr<char []> buffer_uptr = std::make_unique<char []> (io_page_size);
 	  FILEIO_PAGE *second_io_page = reinterpret_cast<FILEIO_PAGE *> (buffer_uptr.get ());
+	  // *INDENT-ON*
 	  error_code = pgbuf_request_data_page_from_page_server (vpid, target_repl_lsa, second_io_page);
 	  if (error_code != NO_ERROR)
 	    {
 	      ASSERT_ERROR ();
 	      return error_code;
 	    }
-	  assert (io_page->prv == second_io_page->prv);
+
+	  /* NOTE:
+	   *  - heap pages are requested also during the recovery phase (after the active transaction server has
+	   *    crashed); in this case, it is normal that the pages requested from page server are out of sync with
+	   *    pages from local storage
+	   * TODO:
+	   *  - in this scenario, what if pages have been deleted, the change is not yet applied on
+	   *    active transaction server (ie: the page still exists), but the log has been applied on
+	   *    page server and the page cannot be retrieved
+	   */
+	  if (log_is_in_crash_recovery_but_past_redo_or_restarted ())
+	    {
+#if !defined(NDEBUG)
+	      if (!(io_page->prv == second_io_page->prv))
+		{
+		  er_log_debug (ARG_FILE_LINE, "pgbuf_read_page_from_file_or_page_server"
+				"past recovery redo pages not equal\n"
+				"    target repl LSA: %lld|%d\n"
+				"    local page  VPID: %d|%d  LSA: %lld|%d  ptype: %d\n"
+				"    remote page VPID: %d|%d  LSA: %lld|%d  ptype: %d\n",
+				LSA_AS_ARGS (&target_repl_lsa),
+				io_page->prv.volid, io_page->prv.pageid,
+				LSA_AS_ARGS (&io_page->prv.lsa), io_page->prv.ptype,
+				second_io_page->prv.volid, second_io_page->prv.pageid,
+				LSA_AS_ARGS (&second_io_page->prv.lsa), second_io_page->prv.ptype);
```

Comment:
How about adding assert (false) here and remove "assert (io_page->prv == second_io_page->prv);"?